R/google_drive.R

```diff
@@ -166,30 +171,50 @@ gd_put <- function(
+  remote_file_changed <- TRUE # assume we'll be changing it
   if(is.na(remote_id)) {
     if(verbose) message("Uploading ", local_file, " to Google Drive")
     remote_id <- googledrive::drive_upload(media=local_file, path=googledrive::as_id(parent), type=type, verbose=verbose)$id
   } else {
     on_exists <- match.arg(on_exists)
-    switch(
-      on_exists,
-      update={
-        if(verbose) message("Updating ", local_file, " on Google Drive")
-        remote_id <- googledrive::drive_update(googledrive::as_id(remote_id), media=local_file, verbose=verbose)$id
-      },
-      replace={
-        if(verbose) message("Replacing ", local_file, " on Google Drive")
-        googledrive::drive_rm(googledrive::as_id(remote_id), verbose=verbose)
-        remote_id <- googledrive::drive_upload(media=local_file, path=googledrive::as_id(parent), type=type, verbose=verbose)$id
-      },
-      stop={
-        stop('File already exists and on_exists==stop')
+    if(on_exists == 'stop') {
+      stop('File already exists and on_exists==stop')
+      # remote_file_changed <- FALSE # true but unnecessary
+    } else {
+      # check Drive to see whether the file we want to post is identical to what's already up there
+      local_hash <- unname(tools::md5sum(data_file))
+      remote_hash <- remote_path %>% slice(n()) %>% pull(drive_resource) %>% .[[1]] %>% .[['md5Checksum']]
+      # if the local file is different from the file on Drive, update or replace it
+      if(local_hash == remote_hash) {
+        if(verbose) message("Not re-posting identical ", local_file, " to Google Drive")
+        remote_file_changed <- FALSE
+      } else {
+        switch(
+          on_exists,
+          update={
+            if(verbose) message("Updating ", local_file, " on Google Drive")
+            remote_id <- googledrive::drive_update(googledrive::as_id(remote_id), media=local_file, verbose=verbose)$id
+          },
+          replace={
+            if(verbose) message("Replacing ", local_file, " on Google Drive")
+            googledrive::drive_rm(googledrive::as_id(remote_id), verbose=verbose)
+            remote_id <- googledrive::drive_upload(media=local_file, path=googledrive::as_id(parent), type=type, verbose=verbose)$id
+          }
+        )
       }
-    )
+    }
   }
   
-  # write the indicator file (involves another check on Google Drive)
-  gd_confirm_posted(ind_file=remote_ind, config_file=config_file)
+  # write the indicator file
+  if(remote_file_changed) {
+    # most common case. involves another check on Google Drive, with patience in
+    # case Drive doesn't fully recognized the new file right away
+    retry_patiently(gd_confirm_posted(ind_file=remote_ind, config_file=config_file), verbose=verbose)
+  } else {
+    # special case. if the remote file was already identical to the local, then
+    # we did not repost the file, and we don't need to check Drive again
+    sc_indicate(ind_file=remote_ind, hash=remote_hash)
```

Comment:
If the remote file is already identical, theoretically the indicator file would also be the same right?

Reply:
I suppose rewriting it wouldn't hurt anything though

Reply:
Yeah...I thought about an additional check so that this was only re-written if the local indicator file didn't exist (which could happen if the analyst manually deleted that file, which he/she might do by accident or if he/she wanted to force this function to run again). However, I also think remake might complain if the target goes untouched...wasn't totally sure about that, but seemed safer to assume that remake _would_ complain and just avoid the issue.